x/sophon/keeper/msg_server_user.go

```diff
@@ -151,12 +151,131 @@ func (m msgServer) RemoveUser(goCtx context.Context, msg *types.MsgRemoveUser) (
 	return &types.MsgRemoveUserResponse{}, nil
 }
 
-func (m msgServer) TopUpUser(_ context.Context, _ *types.MsgTopUpUser) (*types.MsgTopUpUserResponse, error) {
-	panic("not implemented")
+func (m msgServer) TopUpUser(goCtx context.Context, msg *types.MsgTopUpUser) (*types.MsgTopUpUserResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	pubKeyBytes, err := hex.DecodeString(msg.SophonPublicKey)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "invalid hex in pubkey: %s", msg.SophonPublicKey)
+	}
+
+	sophonInfo, err := m.GetSophonInfo(ctx, pubKeyBytes)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return nil, sdkerrors.ErrNotFound.Wrapf("sophon not found for %s", msg.SophonPublicKey)
+		}
+		return nil, err
+	}
+
+	sophonUser, err := m.GetSophonUser(ctx, sophonInfo.Id, msg.UserId)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return nil, sdkerrors.ErrNotFound.Wrapf("user not found for %s", msg.UserId)
+		}
+		return nil, err
+	}
+
+	senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid sender address: %s", msg.Sender)
+	}
+
+	denom, err := m.stakingKeeper.BondDenom(ctx)
+	if err != nil {
+		return nil, err
+	}
+	err = m.bankKeeper.SendCoinsFromAccountToModule(ctx, senderAddr, types.ModuleName, sdk.NewCoins(sdk.NewCoin(denom, msg.Amount)))
+	if err != nil {
+		return nil, err
+	}
+
+	sophonInfo.Balance = sophonInfo.Balance.Add(msg.Amount)
+
+	err = m.SetSophonInfo(ctx, pubKeyBytes, sophonInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	sophonUser.Credits = sophonUser.Credits.Add(msg.Amount)
+
+	err = m.SetSophonUser(ctx, sophonInfo.Id, msg.UserId, sophonUser)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeTopUpUser,
+			sdk.NewAttribute(types.AttributeSophonID, strconv.FormatUint(sophonInfo.Id, 10)),
+			sdk.NewAttribute(types.AttributeSophonPubKey, msg.SophonPublicKey),
+			sdk.NewAttribute(types.AttributeUserID, msg.UserId),
+			sdk.NewAttribute(types.AttributeCredits, msg.Amount.String()),
+		),
+		createUserEvent(msg.SophonPublicKey, sophonInfo.Id, sophonUser),
+		createSophonEvent(sophonInfo),
```

Comment:
Aren't these too many unnecessary attributes not related to this tx especially `createSophonEvent`? And I think there are some duplicates like sophon ID and public key

Reply:
They're separate events though right? My reasoning was if an indexing service only cares about the state of the sophon it can simply monitor `EventTypeSophon` and not have to do manual bookkeeping. That's also why there's duplicate attributes as they're for different events.

Maybe renaming `createSophonEvent` to `newSophonEvent` helps? I meant it as "create a sophon event" and not "event when creating a sophon" :')

Reply:
I'm going to merge the PR but I'm not ignoring the issues :D lets discuss and update the events accordingly in a separate PR, similar to the validation cleanup.